File: cs/src/ConstraintSolver/Intervals/RecursivePropagate.cs
#define USEQUEUE
using System;
using System.IO;
using System.Threading;
using System.Collections.Generic;
//using Alica.Reasoner;
//using Al=Alica;
using AutoDiff;


namespace Alica.Reasoner.IntervalPropagation
{
		
	public class RecursivePropagate : ITermVisitor
	{
		
		//internal Queue<Term> Changed;
		internal TermList Changed;
		
		DownwardPropagator dp = new DownwardPropagator();
		UpwardPropagator up = new UpwardPropagator();
		SetParents sp = new SetParents();
		
		public RecursivePropagate() {
			//this.Changed = new Queue<Term>();
			this.Changed = new TermList();
			dp.Changed = this.Changed;
			up.Changed = this.Changed;
			
		}	
		public void Propagate(Term t) {	
			//for(int i=0; i<2; i++) {
			this.Changed.Clear();
			t.Accept(this);
#if RecPropDEBUG
			Console.WriteLine("Queued Terms: ");
			foreach(Term q in this.Changed) {
					Console.WriteLine(q);					
			}
			Console.WriteLine("------------------------------------");
#endif
			/*
			foreach(Term q in this.Changed) {
				q.Accept(this.sp);
			}
			while(this.Changed.Count > 0) {
				Term cur = this.Changed.Dequeue();
				cur.Accept(this.dp);
				cur.Accept(this.up);				
			}*/
			
			
			Term cur = this.Changed.First;
			while(cur!=null) {
				cur.Accept(this.sp);
				cur = cur.Next;
			}
			
			
			cur =  this.Changed.Dequeue();
			while(cur != null) {
				cur.Accept(this.dp);
				cur.Accept(this.up);
				cur = this.Changed.Dequeue();
			}
		
			//}
			/*cur = this.Changed.First;
			while(cur!=null) {				
				cur.Accept(this.dp);				
				//Term next = cur.Next;
				
				if(cur.Accept(this.up)) {
					/*Term prev = cur.Prev;
					this.Changed.MoveToEnd(cur);					
					if (prev == null) cur = this.Changed.First;
					else cur = prev.Next;*/
					//cur = cur.Next;	
			//	}				
			//	cur = cur.Next;
			//}
			//*/
		}
		private void AddToQueue(Term t) {
			if (!this.Changed.Contains(t)) this.Changed.Enqueue(t);
			//this.Changed.Enqueue(t);
			//this.Changed.MoveToEnd(t);
		}
		
		
        public void Visit(Constant constant) {				
		//	return false;
		}

        public void Visit(Zero zero) {					
		//	return false;
		}

        public void Visit(ConstPower intPower) {			
			AddToQueue(intPower);			
			intPower.Base.Accept(this);
			//return true;
		}
		public void Visit(TermPower intPower) {			
			AddToQueue(intPower);			
			intPower.Base.Accept(this);
			intPower.Exponent.Accept(this);
			//return true;
		}
		public void Visit(Gp gp) {			
			throw new NotImplementedException();
			//return true;
		}
		

        public void Visit(Product product) {
			AddToQueue(product);			
			product.Left.Accept(this);
			product.Right.Accept(this);
			//return true;
		}

        public void Visit(Sigmoid sigmoid) {			
			AddToQueue(sigmoid);
			sigmoid.Arg.Accept(this);
			sigmoid.Mid.Accept(this);
			//return true;
		}
		
		public void Visit(LinSigmoid sigmoid) {			
			AddToQueue(sigmoid);
			sigmoid.Arg.Accept(this);
			//return true;
		}
		
        public void Visit(LTConstraint constraint) {
			AddToQueue(constraint);
			constraint.Left.Accept(this);
			constraint.Right.Accept(this);
			//return true;
		}
		
        public void Visit(LTEConstraint constraint) {
			AddToQueue(constraint);
			constraint.Left.Accept(this);
			constraint.Right.Accept(this);
			//return true;			
		}
		
        public void Visit(Min min) {
			AddToQueue(min);
			min.Left.Accept(this);
			min.Right.Accept(this);
			//return true;
		}
		
        public void Visit(Max max) {
			AddToQueue(max);
			max.Left.Accept(this);
			max.Right.Accept(this);
			//return true;
		}
		
        public void Visit(And and) {
			AddToQueue(and);
			and.Left.Accept(this);
			and.Right.Accept(this);
			//return true;

		}
		
        public void Visit(Or or) {
			AddToQueue(or);
			or.Left.Accept(this);
			or.Right.Accept(this);
			//return true;
		}
		
        public void Visit(ConstraintUtility cu) {
			AddToQueue(cu);
			cu.Constraint.Accept(this);
			cu.Utility.Accept(this);
			//return true;
		}
		
        public void Visit(Sum sum) {
			AddToQueue(sum);
			foreach(Term t in sum.Terms) {
				t.Accept(this);
			}
			//return true;
		}

        public void Visit(AutoDiff.Variable variable) {
			AddToQueue(variable);
			//return true;
		}
		
		public void Visit(Reification reif) {
			AddToQueue(reif);
			reif.Condition.Accept(this);
			//return true;
		}

        public void Visit(Log log) {
			AddToQueue(log);
			log.Arg.Accept(this);
			//return true;			
		}
		
        public void Visit(Sin sin) {
			AddToQueue(sin);
			sin.Arg.Accept(this);
			//return true;						
		}

        public void Visit(Cos cos) {
			AddToQueue(cos);
			cos.Arg.Accept(this);
			//return true;
		}

        public void Visit(Abs abs) {
			AddToQueue(abs);
			abs.Arg.Accept(this);
			//return true;	
		}		
 
        public void Visit(Exp exp) {
			AddToQueue(exp);
			exp.Arg.Accept(this);
			//return true;
		}
		public void Visit(Atan2 atan2) {
			AddToQueue(atan2);
			atan2.Left.Accept(this);
			atan2.Right.Accept(this);
			//return true;
		}
				
		
		
	}
	
}

File: java/src/main/java/Reasoner/IntervalPropagation/RecursivePropagate.java
//
// Translated by CS2J (http://www.cs2j.com): 12.11.2016 09:40:46
//

package Reasoner.IntervalPropagation;

import AutoDiff.Abs;
import AutoDiff.And;
import AutoDiff.Atan2;
import AutoDiff.ConstPower;
import AutoDiff.Constant;
import AutoDiff.ConstraintUtility;
import AutoDiff.Cos;
import AutoDiff.Exp;
import AutoDiff.ITermVisitor;
import AutoDiff.LTConstraint;
import AutoDiff.LTEConstraint;
import AutoDiff.LinSigmoid;
import AutoDiff.Log;
import AutoDiff.Max;
import AutoDiff.Min;
import AutoDiff.Negation;
import AutoDiff.Or;
import AutoDiff.Product;
import AutoDiff.Reification;
import AutoDiff.Sigmoid;
import AutoDiff.Sin;
import AutoDiff.Sum;
import AutoDiff.Term;
import AutoDiff.TermPower;
import AutoDiff.Variable;
import AutoDiff.Zero;

//using Alica.Reasoner;
//using Al=Alica;
public class RecursivePropagate implements ITermVisitor
{
    //internal Queue<Term> Changed;
    public TermList Changed;
    DownwardPropagator dp = new DownwardPropagator();
    UpwardPropagator up = new UpwardPropagator();
    SetParents sp = new SetParents();
    public RecursivePropagate() throws Exception {
        this.Changed = new TermList();
        dp.Changed = this.Changed;
        up.Changed = this.Changed;
    }

    public void propagate(Term t) throws Exception {
        this.Changed.clear();
        t.accept(this);
        Term cur = this.Changed.First;
        while (cur != null)
        {
            cur.accept(this.sp);
            cur = cur.Next;
        }
        cur = this.Changed.dequeue();
        while (cur != null)
        {
            cur.accept(this.dp);
            cur.accept(this.up);
            cur = this.Changed.dequeue();
        }
    }

    private void addToQueue(Term t) throws Exception {
        if (!this.Changed.contains(t))
            this.Changed.enqueue(t);
         
    }

    public Object visit(Constant constant) throws Exception {
        return null;
    }

    //	return false;
    public Object visit(Zero zero) throws Exception {
        return null;
    }

    //	return false;
    public Object visit(ConstPower intPower) throws Exception {
        addToQueue(intPower);
        intPower.getBase().accept(this);
        return null;
    }

    //return true;
    public Object visit(TermPower intPower) throws Exception {
        addToQueue(intPower);
        intPower.getBase().accept(this);
        intPower.getExponent().accept(this);
        return null;
    }


    //return true;
    public Object visit(Product product) throws Exception {
        addToQueue(product);
        product.getLeft().accept(this);
        product.getRight().accept(this);
        return null;
    }

    //return true;
    public Object visit(Sigmoid sigmoid) throws Exception {
        addToQueue(sigmoid);
        sigmoid.getArg().accept(this);
        sigmoid.getMid().accept(this);
        return null;
    }

    //return true;
    public Object visit(LinSigmoid sigmoid) throws Exception {
        addToQueue(sigmoid);
        sigmoid.getArg().accept(this);
        return null;
    }

    //return true;
    public Object visit(LTConstraint constraint) throws Exception {
        addToQueue(constraint);
        constraint.getLeft().accept(this);
        constraint.getRight().accept(this);
        return null;
    }

    //return true;
    public Object visit(LTEConstraint constraint) throws Exception {
        addToQueue(constraint);
        constraint.getLeft().accept(this);
        constraint.getRight().accept(this);
        return null;
    }

    //return true;
    public Object visit(Min min) throws Exception {
        addToQueue(min);
        min.getLeft().accept(this);
        min.getRight().accept(this);
        return null;
    }

    //return true;
    public Object visit(Max max) throws Exception {
        addToQueue(max);
        max.getLeft().accept(this);
        max.getRight().accept(this);
        return null;
    }

    //return true;
    public Object visit(And and) throws Exception {
        addToQueue(and);
        and.getLeft().accept(this);
        and.getRight().accept(this);
        return null;
    }

    //return true;
    public Object visit(Or or) throws Exception {
        addToQueue(or);
        or.getLeft().accept(this);
        or.getRight().accept(this);
        return null;
    }

    //return true;
    public Object visit(ConstraintUtility cu) throws Exception {
        addToQueue(cu);
        cu.getConstraint().accept(this);
        cu.getUtility().accept(this);
        return null;
    }

    //return true;
    public Object visit(Sum sum) throws Exception {
        addToQueue(sum);
        for (Term t : sum.getTerms())
        {
            t.accept(this);
        }
        return null;
    }

    //return true;
    public Object visit(Variable variable) throws Exception {
        addToQueue(variable);
        return null;
    }

    //return true;
    public Object visit(Reification reif) throws Exception {
        addToQueue(reif);
        reif.getCondition().accept(this);
        return null;
    }

    @Override
    public Object visit(final Negation r) throws Exception
    {
        return null;
    }

    //return true;
    public Object visit(Log log) throws Exception {
        addToQueue(log);
        log.getArg().accept(this);
        return null;
    }

    //return true;
    public Object visit(Sin sin) throws Exception {
        addToQueue(sin);
        sin.getArg().accept(this);
        return null;
    }

    //return true;
    public Object visit(Cos cos) throws Exception {
        addToQueue(cos);
        cos.getArg().accept(this);
        return null;
    }

    //return true;
    public Object visit(Abs abs) throws Exception {
        addToQueue(abs);
        abs.getArg().accept(this);
        return null;
    }

    //return true;
    public Object visit(Exp exp) throws Exception {
        addToQueue(exp);
        exp.getArg().accept(this);
        return null;
    }

    //return true;
    public Object visit(Atan2 atan2) throws Exception {
        addToQueue(atan2);
        atan2.getLeft().accept(this);
        atan2.getRight().accept(this);
        return null;
    }

}


//return true;